torch/csrc/autograd/engine.cpp

```diff
@@ -228,14 +237,19 @@ static variable_list call_post_hooks(Function& fn, variable_list outputs, variab
 }
 
 static variable_list call_function(FunctionTask& task) {
+  prev_checkpoint_valid_state = checkpoint_valid;
+  checkpoint_valid = (!task.base->checkpoint_grad_inputs) && prev_checkpoint_valid_state;
   auto& fn = *task.fn;
   auto inputs = call_pre_hooks(fn, InputBuffer::variables(std::move(task.inputs)));
 
   if(!task.base->keep_graph) {
     fn.will_release_variables();
   }
   auto outputs = fn(inputs);
-
+  // if (task.base->checkpoint_grad_inputs) {
```

Comment:
@ezyang , I feel like restoring state is not really needed. Imagine this:

.backward() -> .grad() -> .backward()

in this case, checkpoint_valid pattern would be:  True -> False -> False
If we restore the checkpoint_valid prev_checkpoint_valid_state would be: True-> True -> True which is incorrect

what do you think?